### PR TITLE
feat: Add suport for `Concurrent::Promises::Future`

### PR DESCRIPTION
**Pull request by miyucy**
Hi, I'm using concurrent-ruby gem in my rails projects.
but, `opentelemetry-instrumentation-concurrent_ruby` couldn't instrument `Concurrent::Promises::Future` class correctly.

**Share details about your runtime**

Operating system details: macOS, Darwin 20.6.0
RUBY_ENGINE: "ruby"
RUBY_VERSION: "2.7.2"
RUBY_DESCRIPTION: "ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-darwin20]"

**Share a simplified reproduction if possible**

```rb
require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  gem 'opentelemetry-api'
  gem 'opentelemetry-sdk'
  gem 'opentelemetry-instrumentation-concurrent_ruby'
  gem 'concurrent-ruby'
end

require 'opentelemetry-api'
require 'opentelemetry-sdk'
require 'opentelemetry-instrumentation-concurrent_ruby'

exporter = OpenTelemetry::SDK::Trace::Export::InMemorySpanExporter.new
span_processor = OpenTelemetry::SDK::Trace::Export::SimpleSpanProcessor.new(
  exporter
)

OpenTelemetry::SDK.configure do |c|
  c.use 'OpenTelemetry::Instrumentation::ConcurrentRuby'
  c.add_span_processor(span_processor)
end

SimpleTracer = OpenTelemetry.tracer_provider.tracer('Bug Report')

SimpleTracer.in_span('Parent span') do
  future = Concurrent::Promises.future do
    SimpleTracer.in_span('child span') {}
  end
  future.value
end

exporter.finished_spans.size == 2
# => true
parent_span = exporter.finished_spans.find { |span| span.name == 'Parent span' }
child_span = exporter.finished_spans.find { |span| span.name == 'child span' }
p child_span.parent_span_id == parent_span.span_id
# => false. but, should be true

OpenTelemetry.tracer_provider.shutdown
```

Related issue: https://github.com/DataDog/dd-trace-rb/issues/1316

### CODE REVIEW

**Comment by arielvalentin** [comment]
@open-telemetry/ruby-maintainers What are our current guidelines for version comparability?

**Comment by ericmustin** [comment]
@miyucy hello! thanks for the PR here. I have this on my plate to followup on, but admittedly am a bit rusty on concurrent_ruby, so let me review how this is handled in some other instrumentation gems i'm a bit more familiar with. should hopefully have some next steps here in the next day or so. 

also good question @arielvalentin . let me try to organize my 2 cents here. probably something worth formalizing and documenting once we reach consensus.

**Comment by arielvalentin** [comment]
@ericmustin @miyucy There are a few things that stand out to me in this PR. 

### Version Compatibility - What versions of `concurrent_ruby` should this support?

I took a look at instrumentation base and I can see that there is a [`compatible`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/instrumentation/base/lib/opentelemetry/instrumentation/base.rb#L136) method use to guard against incompatible versions of gems. 

It looks like in [most cases](https://github.com/open-telemetry/opentelemetry-ruby/search?q=%22compatible+do%22) the implementation checks that there is at least a minimum gem version, however there isn't one present for [`concurrent_ruby`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb). I think we should add some sort of version compatibility check before merging this PR.

This PR drops support for the deprecated `Concurrent::Future` class, however it is still present in [v1.1.9](https://github.com/ruby-concurrency/concurrent-ruby/blob/v1.1.9/lib/concurrent-ruby/concurrent/future.rb). Should we continue to support this deprecated class until it is removed for cases where users have not yet migrated towards Promises? i.e. keep both implementations until it is removed in a future release?

### Platform Support

This change also introduces instrumentation for the Ruby Implementation of the Executor Service however `concurrent_ruby` also supports JRuby, which it determines to load at [runtime](https://github.com/ruby-concurrency/concurrent-ruby/blob/7b7900a9133a5c31e60fc55133559ae07c934d69/lib/concurrent-ruby/concurrent/executor/thread_pool_executor.rb#L10). 

In order to support both platforms, should the [`post`](https://github.com/open-telemetry/opentelemetry-ruby/pull/943/files#diff-4625ee4ece70ed3b76518b3373b35a44425a4c86e4404b3d5b01ee452ce2cf78R14) override get mix into the public interface [`ThreadPoolExector`](https://github.com/ruby-concurrency/concurrent-ruby/blob/7b7900a9133a5c31e60fc55133559ae07c934d69/lib/concurrent-ruby/concurrent/executor/thread_pool_executor.rb#L56) or do plan on only supporting Ruby at this time?

**Comment by miyucy** [comment]
@arielvalentin Thank you for review!

> This PR drops support for the deprecated `Concurrent::Future` class

No.  this PR still support `Concurrent::Future`.  You can see how it work in 
 example file (`instrumentation/concurrent_ruby/example/concurrent_ruby.rb`).

> Platform Support - Ruby && JRuby

Sorry, I did not try to other ruby implementations... But, I think this library should support other ruby impls.

> should the post override get mix into the public interface ThreadPoolExector.

Thank you for impl suggestion. I will try to change.

**Comment by arielvalentin** [comment]
@miyucy Thank you for 

> @arielvalentin Thank you for review!
> 
> > This PR drops support for the deprecated `Concurrent::Future` class
> 
> No. this PR still support `Concurrent::Future`. You can see how it work in
> example file (`instrumentation/concurrent_ruby/example/concurrent_ruby.rb`).
> 

My mistake! I am very sorry that I missed that, I did see that the example was unchanged from the diff and after looking more closely at the PR the test for `Concurrent::Future` are still there. 

Thank you for preserving the functionality. 

> > Platform Support - Ruby && JRuby
> 
> Sorry, I did not try to other ruby implementations... But, I think this library should support other ruby impls.
> 
> > should the post override get mix into the public interface ThreadPoolExector.
> 
> Thank you for impl suggestion. I will try to change.

Thank you very much.

**Comment by miyucy** [comment]
@arielvalentin Hi, I just updated PR. I confirmed my patches works well ruby and jruby. Could you review again?

```
app@d6ba093fe046:/app/instrumentation/concurrent_ruby/example$ bundle exec ruby -v concurrent_ruby.rb 

jruby 9.2.19.0 (2.5.8) 2021-06-15 55810c552b OpenJDK 64-Bit Server VM 25.302-b08 on 1.8.0_302-b08 +jit [linux-x86_64]
I, [2021-09-17T06:39:11.660810 #109]  INFO -- : Instrumentation: OpenTelemetry::Instrumentation::ConcurrentRuby was successfully installed
#<struct OpenTelemetry::SDK::Trace::SpanData
 name="inner_span",
 kind=nil,
 status=#<OpenTelemetry::Trace::Status:0x43795f04 @code=1, @description="">,
 parent_span_id="Y\xE7!\xE2\xFCQ\xBF\xB3",
 total_recorded_attributes=0,
 total_recorded_events=0,
 total_recorded_links=0,
 start_timestamp=1631860751703836000,
 end_timestamp=1631860751704193000,
 attributes={},
 links=nil,
 events=nil,
 resource=
  #<OpenTelemetry::SDK::Resources::Resource:0x6bb9b5c5
   @attributes=
    {"service.name"=>"unknown_service",
     "process.pid"=>109,
     "process.command"=>"concurrent_ruby.rb",
     "process.runtime.name"=>"jruby",
     "process.runtime.version"=>"2.5.8",
     "process.runtime.description"=>
      "jruby 9.2.19.0 (2.5.8) 2021-06-15 55810c552b OpenJDK 64-Bit Server VM 25.302-b08 on 1.8.0_302-b08 +jit [linux-x86_64]",
     "telemetry.sdk.name"=>"opentelemetry",
     "telemetry.sdk.language"=>"ruby",
     "telemetry.sdk.version"=>"1.0.0.rc3"}>,
 instrumentation_library=
  #<struct OpenTelemetry::SDK::InstrumentationLibrary
   name="default",
   version="">,
 span_id="\x05\xE9w!\x97_X;",
 trace_id="\xB73\x17\x01\xB4\xA4\xF91\xAFG\xF3\xF7\xC3{%E",
 trace_flags=#<OpenTelemetry::Trace::TraceFlags:0x4ee462a9 @flags=1>,
 tracestate=#<OpenTelemetry::Trace::Tracestate:0x27739af3 @hash={}>>
#<struct OpenTelemetry::SDK::Trace::SpanData
 name="outer_span",
 kind=nil,
 status=#<OpenTelemetry::Trace::Status:0x43795f04 @code=1, @description="">,
 parent_span_id="\x00\x00\x00\x00\x00\x00\x00\x00",
 total_recorded_attributes=0,
 total_recorded_events=0,
 total_recorded_links=0,
 start_timestamp=1631860751671142000,
 end_timestamp=1631860751833472000,
 attributes={},
 links=nil,
 events=nil,
 resource=
  #<OpenTelemetry::SDK::Resources::Resource:0x6bb9b5c5
   @attributes=
    {"service.name"=>"unknown_service",
     "process.pid"=>109,
     "process.command"=>"concurrent_ruby.rb",
     "process.runtime.name"=>"jruby",
     "process.runtime.version"=>"2.5.8",
     "process.runtime.description"=>
      "jruby 9.2.19.0 (2.5.8) 2021-06-15 55810c552b OpenJDK 64-Bit Server VM 25.302-b08 on 1.8.0_302-b08 +jit [linux-x86_64]",
     "telemetry.sdk.name"=>"opentelemetry",
     "telemetry.sdk.language"=>"ruby",
     "telemetry.sdk.version"=>"1.0.0.rc3"}>,
 instrumentation_library=
  #<struct OpenTelemetry::SDK::InstrumentationLibrary
   name="default",
   version="">,
 span_id="Y\xE7!\xE2\xFCQ\xBF\xB3",
 trace_id="\xB73\x17\x01\xB4\xA4\xF91\xAFG\xF3\xF7\xC3{%E",
 trace_flags=#<OpenTelemetry::Trace::TraceFlags:0x4ee462a9 @flags=1>,
 tracestate=#<OpenTelemetry::Trace::Tracestate:0x27739af3 @hash={}>>
#<struct OpenTelemetry::SDK::Trace::SpanData
 name="inner_span",
 kind=nil,
 status=#<OpenTelemetry::Trace::Status:0x43795f04 @code=1, @description="">,
 parent_span_id="I\e\x827\xDF\t%\x81",
 total_recorded_attributes=0,
 total_recorded_events=0,
 total_recorded_links=0,
 start_timestamp=1631860751955331000,
 end_timestamp=1631860751955780000,
 attributes={},
 links=nil,
 events=nil,
 resource=
  #<OpenTelemetry::SDK::Resources::Resource:0x6bb9b5c5
   @attributes=
    {"service.name"=>"unknown_service",
     "process.pid"=>109,
     "process.command"=>"concurrent_ruby.rb",
     "process.runtime.name"=>"jruby",
     "process.runtime.version"=>"2.5.8",
     "process.runtime.description"=>
      "jruby 9.2.19.0 (2.5.8) 2021-06-15 55810c552b OpenJDK 64-Bit Server VM 25.302-b08 on 1.8.0_302-b08 +jit [linux-x86_64]",
     "telemetry.sdk.name"=>"opentelemetry",
     "telemetry.sdk.language"=>"ruby",
     "telemetry.sdk.version"=>"1.0.0.rc3"}>,
 instrumentation_library=
  #<struct OpenTelemetry::SDK::InstrumentationLibrary
   name="default",
   version="">,
 span_id="b\xE6\xCF\xB6\xE6o g",
 trace_id="f\xD9<LZ\x86N\xF7\x94H\xC3\xDB\xAE\xAB\xADT",
 trace_flags=#<OpenTelemetry::Trace::TraceFlags:0x4ee462a9 @flags=1>,
 tracestate=#<OpenTelemetry::Trace::Tracestate:0x27739af3 @hash={}>>
#<struct OpenTelemetry::SDK::Trace::SpanData
 name="outer_span",
 kind=nil,
 status=#<OpenTelemetry::Trace::Status:0x43795f04 @code=1, @description="">,
 parent_span_id="\x00\x00\x00\x00\x00\x00\x00\x00",
 total_recorded_attributes=0,
 total_recorded_events=0,
 total_recorded_links=0,
 start_timestamp=1631860751949977000,
 end_timestamp=1631860752001643000,
 attributes={},
 links=nil,
 events=nil,
 resource=
  #<OpenTelemetry::SDK::Resources::Resource:0x6bb9b5c5
   @attributes=
    {"service.name"=>"unknown_service",
     "process.pid"=>109,
     "process.command"=>"concurrent_ruby.rb",
     "process.runtime.name"=>"jruby",
     "process.runtime.version"=>"2.5.8",
     "process.runtime.description"=>
      "jruby 9.2.19.0 (2.5.8) 2021-06-15 55810c552b OpenJDK 64-Bit Server VM 25.302-b08 on 1.8.0_302-b08 +jit [linux-x86_64]",
     "telemetry.sdk.name"=>"opentelemetry",
     "telemetry.sdk.language"=>"ruby",
     "telemetry.sdk.version"=>"1.0.0.rc3"}>,
 instrumentation_library=
  #<struct OpenTelemetry::SDK::InstrumentationLibrary
   name="default",
   version="">,
 span_id="I\e\x827\xDF\t%\x81",
 trace_id="f\xD9<LZ\x86N\xF7\x94H\xC3\xDB\xAE\xAB\xADT",
 trace_flags=#<OpenTelemetry::Trace::TraceFlags:0x4ee462a9 @flags=1>,
 tracestate=#<OpenTelemetry::Trace::Tracestate:0x27739af3 @hash={}>>
```

File: instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb
Comment (arielvalentin) [inline instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb:29]: Could you help me understand this decision here? 
Would it work if this was mixed into `ThreadPoolExectutor` instead?

Comment (miyucy) [reply instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb:29]: Because, some classes inherit from `RubyExecutorService`. So I mixed into that class.
(But, I don't know Concurrent::TimerTask works at JRuby environment...)

File: instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb
Comment (miyucy) [inline instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb:33]:
```suggestion
          ::Concurrent::RubyExecutorService.prepend Patches::ExecutorService          
          ::Concurrent::JavaExecutorService.prepend Patches::ExecutorService if ::Concurrent.on_jruby?
```

perhaps, should I change like above?

File: instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb
Comment (arielvalentin) [reply instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb:29]: Thank you for taking the time to explain that. 

Looking at other API docs it looks like there are a few other Ruby specific implementations so this does indeed seem necessary.
- https://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/SimpleExecutorService.html 
- https://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/TimerSet.html

Comment (miyucy) [reply instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb]: Thank you for review. I'll verify that class works at JRuby environment later.

Comment (arielvalentin) [reply instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb]: 🤔 @ericmustin If this is prepended only on the platform specific executor services, then the derived classes may not be instrumented the way we expect. I think I am a little confused by some of the implementations in the derived classes.

`TimerSet` overrides `post` and does not invoke `super` so it will not be auto-instrumented as a result, but that may be ok?

https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent-ruby/concurrent/executor/timer_set.rb#L48

While the `TimerTask` _changes_ the interface by encapsulating the `post` method altogether and introduces a new interface named `execute`, which will also not be auto-instrumented: 

https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent-ruby/concurrent/timer_task.rb#L273

`SimpleExecutorService` _also_ overrides `post` and does not invoke `super`:
https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent-ruby/concurrent/executor/simple_executor_service.rb#L37

❓ Do we want to instrument any of these derived classes? If so then I _think_ it would be outside of the scope of this PR, but that goes back to me wanting this PR to patch the non-platform specific `ThreadPoolExectutor` and then look more closely at the other Ruby platform specific classes to see what/how to auto-instrument those in a future PR.

Comment (miyucy) [reply instrumentation/concurrent_ruby/lib/opentelemetry/instrumentation/concurrent_ruby/instrumentation.rb]: @arielvalentin 

Hi, I tried below codes MRI and JRuby. But, it couldnot instrument as expected. because `Concurrent::ScheduledTask` does not use concurrent executor service.

```ruby
tracer.in_span('outer_span') do
  task = Concurrent::ScheduledTask.execute(0.5) do
    p Thread.current.object_id
    tracer.in_span('inner_span') {
      p Thread.current.object_id
      puts "Concurrent::ScheduledTask"
    }
  end
  p task.state
end
sleep 1 while task.pending?
p task.state

exporter.finished_spans.each_with_index do |span, i|
  puts "i=#{i} name=#{span.name} span_id=#{span.span_id.inspect} parent_span_id=#{span.parent_span_id.inspect}"
end
```

> wanting this PR to patch the non-platform specific ThreadPoolExectutor and then look more 

So I changing my mind. agree with your option.

**Review by fbogsany — APPROVED** [review]
The LGTM, but I'd really like to get some more 👀 on this before I merge.

**Review by arielvalentin — APPROVED** [review]
Thanks for your contribution! 🎉

**Pull request merged**
